DRAFT: Added additional external resources doc content (#363)
okteto/docs: src/content/cloud/external-resources.mdx

---

> title: External Resources

Reviewer: I am missing a comment about the common use case where:
- External resource is deployed in your `deploy` commands
- External resource is destroyed in your `destroy` commands
- You can configure the `deploy/destroy` image
- External resources are deleted when the namespace is deleted. This in combination with the Okteto garbage collector makes sure you never leave cloud resources hanginf for ever.

Author: @pchico83 Do we have a sample `deploy` or `destroy` config we could include here?

Or is the broader point your one from our call today where you're saying the external resources can be included as part of a new image and that image can be configured in the `deploy` and `destroy` sections?

Reviewer: @omnomagonz external resources let you manage the life cycle of cloud external resources as part of your development environment. In addition to the option to customize endpoints and notes.txt in the UI, you can configure your own container image to run your deploy/destroy commands the same way for all your developers.
@rberrelleza is working on samples. @rberrelleza could you provide a good sample here?

Author: Hopefully I addressed these gaps in the latest commit 🤞

---
title: External Resources

> description: Configure Okteto's manifest to show external resources in your development environment

Reviewer 2: ```suggestion
description: Configure Okteto's manifest to include external resources in your development environment
```

sidebar_label: External Resources
id: external-resources
---

import Image from '@theme/Image';

> The External Resources feature in Okteto is a powerful mechanism that enables you to manage the lifecycle of cloud external resources as part of your Okteto development environment. This means you can define the external resources that comprise your application, include those resources in Okteto's garbage collection process, and even destroy those resources when you finish working with your development environment.

Reviewer 2: Can we add some examples of external resources? 

```suggestion
The External Resources feature in Okteto is a powerful mechanism that enables you to manage the lifecycle of external resources as part of your Okteto development environment. This means you can define the external resources that comprise your application, include those resources in Okteto's garbage collection process, and even destroy those resources when you finish working with your development environment.
```

You no longer need to use different tools to work on your application in Okteto and you won't need to worry about rogue external resources continuing to run once you are done with your work.

When you configure External Resources in Okteto you can see those resources within the Okteto UI, can click to go directly to those resources in their respective platform, and include them in your Okteto `deploy` and `destroy` configurations.

## Overview

When you configure the `external` section in your [Okteto manifest](/reference/manifest), keep in mind the following:

> - You can create an image that includes your desired external (to Okteto) resources you'd like to use for your application

Reviewer 2: ```suggestion
- You can create a container image that includes any external tools (to Okteto) that you need in order to provision or access your external resources.
```

> - This image can be included in your Okteto manifest's `deploy` and `destroy` sections

Reviewer 2: ```suggestion
- This container image can be included in your Okteto manifest's `deploy` and `destroy` sections
```

- External resources are deleted when your Okteto namespace is deleted
- External resources are covered by Okteto's garbage collection process

## Configuring the `external` section

Let's take a look at the composition of the `external` section in the Okteto manifest starting with the high-level structure of this configuration block. There are some important details you need to know before using this feature.

Here's an example of how to define your `external` resources in the Okteto manifest, which you can see in our [References section](/reference/manifest/#external-object-optional):

```yaml

> external:

Reviewer 2: The `external` section is to tell okteto which URL to display. The person defining the resource must also include the necessary steps in `deploy` to provision the resource (e.g. call the AWS API to create an s3 bucket). We don't automate that part today, but we will make samples and building blocks available.

Author: Ahhh, that's great clarification. Let me rework this section a bit 👍

Author: @rberrelleza Do we automate the `destroy` action or does an external resource also need to be configured there to be destroyed?

Author: [Latest commit](https://github.com/okteto/docs/pull/363/files#diff-b5fcc1ba2f90c8eced81c0c31ce915bb44abc500bc7e6f24c7a4a998216e4769R35) attempts to expand on these dynamics.

Let me know if that works 🙏

  db:
    notes: docs/database.md
    icon: database
    endpoints:
    - name: db
      url: https://localhost:3306
  functions:
    notes: docs/lambdas.md
    icon: function
    endpoints:
    - name: data-aggregator
      url: https://fake-id.lambda-url.us-east-1.on.aws.aggregator
    - name: data-processor
      url: https://fake-id.lambda-url.us-east-1.on.aws.processor
```

> You can also use the [dynamic endpoints](/reference/manifest/#use-dynamic-endpoints-for-an-external-url) feature here to provide more flexibility in the URLs.

Every keyword that you see here is opinionated except for `db:` and `functions:`. Those two keywords, the names of the external resources being defined here, can be anything you want them to be. For example, if we want to call the `db:` resource `dundie_award:` you can do so. Similarly, you could name `functions:` something like `parkour:` and it will work just fine.

Whatever name you use to label these resources will be what's shown in the Okteto UI. Choose wisely!

## Defining Your External Resources

As you configure your external resources in this section, keep in mind the following definitions:

> `notes:` is used to provide context in the form of a relative link to a markdown file. The contents are displayed in the Okteto UI (see below).

Reviewer 2: The link must point to a file that lives on the repository. Also, we have some rules on the markdown we support. should we put them here?

Author: @rberrelleza Yes, I think we should include those rules. Are they documented somewhere that you know of?

<p align="center"><Image src="/docs/cloud/images/external_resource_notes.png" alt="External resources notes section" width="450" /></p>

`icon:` specifies one of several predefined icons you can use when this resource is rendered in the Okteto UI. The current list of supported icons is:

- container
- dashboard
- database
- default
- function
- graph
- storage

<p align="center"><Image src="/docs/cloud/images/external_resource_list.png" alt="External resources listed in dev environment" width="450" /></p>

`endpoints:` contain information on how to access the external resource using:

- `name:` to name the endpoint you're defining
- `url:` to provide the specific url for accessing the resource

> The name of the endpoint is used to inject or modify environment variables. In this example, one of those would be: `OKTETO_EXTERNAL_FUNCTIONS_ENDPOINTS_DATA_AGGREGATOR_URL`.

Reviewer 2: Can we add a section on dynamic endpoints? It's something that everyone will need for non-trivial samples.

Author: You bet.

I'll take a stab at it and we can review 👍

Author: Latest commit [takes a pass](https://github.com/okteto/docs/pull/363/files#diff-b5fcc1ba2f90c8eced81c0c31ce915bb44abc500bc7e6f24c7a4a998216e4769R127) at this. Let me know what you think 😄

All of this information can be found and in a more concise format in our [References section](/reference/manifest/#external-object-optional).

## Example: Deploy a Development Environment with External Resources

> Now that we've walked through a sample of an `external` configuration, let's look at using this within the broader context of an application we're working on. You can navigate to our [demo repo](https://github.com/okteto/voting-app-with-external-resources) to follow along if you'd like.

Reviewer 2: can we use https://github.com/okteto/external-resources-aws instead? I think AWS resources might be better for this, but what do you think?

Author: @rberrelleza The impression I've been getting from you is that this `external-resources-aws` demo is the more formal one that will also be part of our GTM. I think it makes sense to include that instead 👍

Author: Latest commit uses this new demo repo and I think I've updated all of the relevant language. Feel free to review again when able

Here are just a few examples of external resources you can use with Okteto:

- A markdown file (e.g. README, API Documentation)
- MongoDB Atlas Instance
- LaunchDarkly
- Amazon SQS
- Amazon S3
- Amazon RDS databases
- AWS Lambda Functions

You no longer need to use different tools to work on your application in Okteto and you won't need to worry about rogue external resources continuing to run once you are done with your work.

When you configure External Resources in Okteto you can see those resources within the Okteto UI, can click to go directly to those resources in their respective platform, and include them in your Okteto `deploy` and `destroy` configurations.

## Overview

When you configure the `external` section in your [Okteto manifest](/reference/manifest), keep in mind the following:

- You can create a container image that includes any external tools (to Okteto) that you need in order to provision or access your external resources.
- This container image can be included in your Okteto manifest's `deploy` and `destroy` sections
- External resources are deleted when your Okteto namespace is deleted

> - External resources are covered by Okteto's garbage collection process

Reviewer: Should we link https://www.okteto.com/docs/administration/cleanup/ here?

Author: Latest commit adds that link 👍

## Configuring external resources

Using external resources with Okteto requires configuring three areas of your Okteto manifest:

1. The `deploy` section to provision your external resources (e.g. call the AWS API to create an s3 bucket)
2. The `destroy` section to ensure the external resource is destroyed (as appropriate) when your development environment is destroyed
3. The `external` section to populate the Okteto UI and link to the external resource

### Configuring the `deploy` section

When you use an external resource with Okteto it needs to be provisioned like any other resource. If we use the example above of creating an AWS S3 bucket, we could add the following to our Okteto manifest in the `deploy` section:

```yaml
deploy:
  image: okteto/pipeline-runner:1.0.0-sam
  commands:  
  - name: Create an AWS S3 bucket
    command: |
      sh s3/deploy.sh
```

> In this example we're using a script that contains all of the instructions to do the provisioning work and executing that script in this particular stage of the Okteto `deploy` section. 

Author: Latest commit adds a note to mention this 👍

---
title: External Resources
description: Configure Okteto's manifest to include external resources in your development environment
sidebar_label: External Resources
id: external-resources
---

import Image from '@theme/Image';

The External Resources feature in Okteto is a powerful mechanism that enables you to manage the lifecycle of external resources as part of your Okteto development environment. This means you can define the external resources that comprise your application, include those resources in Okteto's garbage collection process, and even destroy those resources when you finish working with your development environment. 

Here are just a few examples of external resources you can use with Okteto:

- A markdown file (e.g. README, API Documentation)
- MongoDB Atlas Instance
- LaunchDarkly
- Amazon SQS
- Amazon S3
- Amazon RDS databases

> - AWS Lambda Functions

Author: Added additional samples and language to emphasize the flexibility here

You no longer need to use different tools to work on your application in Okteto and you won't need to worry about rogue external resources continuing to run once you are done with your work.

When you configure External Resources in Okteto you can see those resources within the Okteto UI, can click to go directly to those resources in their respective platform, and include them in your Okteto `deploy` and `destroy` configurations.

## Overview

When you configure the `external` section in your [Okteto manifest](/reference/manifest), keep in mind the following:

- You can create a container image that includes any external tools (to Okteto) that you need in order to provision or access your external resources.
- This container image can be included in your Okteto manifest's `deploy` and `destroy` sections
- External resources are destroyed when your Okteto namespace is deleted
- External resources are covered by Okteto's [garbage collection process](/administration/cleanup/#delete-unused-namespaces)

> ## Configuring external resources

Reviewer 2: The keys to access the resources also need to be set. Should we mention this here? (Or do you plan an admin-level counterpart for this?

This is a big value add: admin centralizes credentials instead if giving them to every developer

Author: @rberrelleza I hadn't planned a separate section. I think it makes sense to include here and we can consider splitting it later if we find that's the right move 😄 

For my edification: are you saying that an `admin` could (not a hard requirement) specify the `Secrets` in the `Admin` > `Secrets` page that then all developers would use for a given resource? 

So all namespaces that deploy an application that uses (for example) AWS S3 buckets would rely on the admin-set keys for an organization in AWS vs creating their own AWS accounts and managing those secrets locally on their own Okteto account?

Just want to calibrate so I know how to write this section 👍

Reviewer 2: Exactly. , that's the way to go for anything that requires an API key.

For AWS and GCP specifically, we also support a "credential-less" model for self-hosted customers, where they can define security roles and use those to give permissions to Okteto instead of using AP Keys. This is the recommended way by both AWS and GCP (Provecho is giving a talk about this on KubeCon if you are interested).

Author: Latest commit includes some language to address this: https://github.com/okteto/docs/pull/363/commits/7f3d39ef99fe1ab10b55992f331083da4524d11b

Using external resources with Okteto requires configuring three areas of your Okteto manifest:

1. The `deploy` section to provision your external resources (e.g. call the AWS API to create an s3 bucket)
2. The `destroy` section to ensure the external resource is destroyed (as appropriate) when your development environment is destroyed
3. The `external` section to populate the Okteto UI and link to the external resource

### Configuring the `deploy` section

When you use an external resource with Okteto it needs to be provisioned like any other resource. If we use the example above of creating an AWS S3 bucket, we could add the following to our Okteto manifest in the `deploy` section:

```yaml
deploy:
  image: okteto/pipeline-runner:1.0.0-sam
  commands:  
  - name: Create an AWS S3 bucket
    command: |

>       sh s3/deploy.sh

Author: In the latest commit (58b9d8f90c5afaaaf3ce9eadc19edf63560a3a96) I linked to the demo repo's `deploy.sh` script so they can explore if they'd like. The way it's written now seems cleaner and the new link should help clarify 🤞

```

That will compile the appropriate URL, using Okteto environment variables, that specify where a developer in the same namespace will go when they click on the AWS SQS external resource component in the Okteto UI (see below).

<p align="center"><Image src="/docs/cloud/images/external_resources_sqs_demo.png" alt="AWS SQS external resource demo UI component" width="450" /></p>

## Example: Deploy a Development Environment with External Resources

Now that we've walked through a sample of an `external` configuration, let's look at using this within the broader context of an application we're working on. You can navigate to our [demo repo](https://github.com/okteto/external-resources-aws) to follow along if you'd like.

In this example we'll be configuring and deploying a development environment that includes polyglot microservices, an AWS SQS queue, and an S3 bucket.

<p align="center"><Image src="/docs/cloud/images/external_resources_architecture.png" alt="Application architecture diagram" width="450" /></p>

### Prerequisites

1. Okteto CLI 2.14 or newer
1. An AWS account
1. An Okteto account

> 1. Create a set of IAM keys for your AWS account (If you are using Okteto Self-Hosted, you can directly assign an AWS Role)

Reviewer 2: Link to how to configure the role?

Author: Added in latest commit: 58b9d8f90c5afaaaf3ce9eadc19edf63560a3a96

Feel free to double check the links are the ones you had in mind 😅

> 1. Create the following Okteto secrets:

Reviewer 2: Link to secrets?

Author: Added in latest commit: 58b9d8f90c5afaaaf3ce9eadc19edf63560a3a96

```yaml
external:
  readme:
    icon: okteto
    notes: README.md
    endpoints:
    - name: readme
      url: https://github.com/okteto/external-resources-aws
  sqs:
    icon: aws
    notes: sqs/notes.md
    endpoints:
    - name: queue
  s3:
    icon: aws
    notes: s3/notes.md
    endpoints:
    - name: bucket
  api-docs:
    icon: dashboard
    notes: check/notes.md
    endpoints:
    - name: docs
```

The `url:` keys are missing from most external resource (`sqs`, `s3`, and `api-docs`) because they've been defined in the `deploy` > `commands` > `command` section using the [Okteto Secrets](/cloud/secrets) you defined [above](#prerequisites) and the [dynamic endpoints](#using-dynamic-endpoints-with-external-resources) feature.

Alternatively, you could remove the respective `deploy` commands and add the `url` endpoints here manually, but this is a good example of how flexible the manifest file can be.

Once you've configured the prerequisites, you can deploy this on your Okteto instance:

```
$ git clone https://github.com/okteto/external-resources-aws
$ cd external-resources-aws
$ okteto context use $OKTETO_URL
$ okteto deploy
```

> Once the application has been deployed in Okteto, you can work on the `menu` microservice by running the command: `okteto up menu`. In this specific example, once you've executed this command you'll also need to run `yarn start` from the container.

Reviewer 2: Can we add a paragraph iterating that the command above will deploy the environment and the cloud resources and all the things?

> Once the application has been deployed in Okteto, you can work on the `menu` microservice by running the command: `okteto up menu`. In this specific example, once you've executed this command you'll also need to run `yarn start` from the container.

Reviewer 2: Can we add a paragraph iterating that the command above will deploy the environment and the cloud resources and all the things?
